File: NeuralNet01/Net.cpp
#include "Net.h"
#include "Neuron.h"
#include <cassert>
#include <iostream>

Net::Net(const std::vector<unsigned int> topology)
{
	unsigned numLayers = topology.size();
	for (unsigned int layerNum = 0; layerNum < numLayers; ++layerNum)
	{
		unsigned int numOutputs = 0;
		if (layerNum != topology.size() - 1)
			numOutputs = topology[layerNum + 1];

		m_Layers.push_back(Layer());
		for (unsigned int neuronNum = 0; neuronNum <= topology[layerNum]; ++neuronNum)//1 bias neuron so 1 more loop
		{
			m_Layers.back().push_back(Neuron(numOutputs, neuronNum));//back gives most recent element
		}

		//force the bias neuron to output 1
		m_Layers.back().back().SetOutputValue(1.0);
	}
}

void Net::getResults(std::vector<double>& resultsVals) const
{
	resultsVals.clear();
	for (unsigned int n = 0; n < m_Layers.back().size() - 1; ++n)
	{
		resultsVals.push_back(m_Layers.back()[n].GetOutputValue());
	}
}

void Net::backProp(const std::vector<double>& targetVals)
{
	// calculate overall net error (RMS of output neuron errors)
	//RMS = root mean square
	Layer &outputLayer = m_Layers.back();
	m_error = 0.0;

	for (unsigned int n = 0; n < outputLayer.size() - 1; ++n)
	{
		double delta = targetVals[n] - outputLayer[n].GetOutputValue();
		m_error += delta * delta;
	}
	m_error /= outputLayer.size() - 1;
	m_error = sqrt(m_error);//rms

	//recent average measurement:
	m_recentAverageError = (m_recentAverageError * m_recentAverageSmoothingFactor + m_error) / (m_recentAverageSmoothingFactor + 1.0);

	//calculate output layer gradients
	for (unsigned int n = 0; n < outputLayer.size() - 1; ++n)
	{
		outputLayer[n].calcOutputGradients(targetVals[n]);
	}

	//calculate gradients on hidden layers
	//for (unsigned int layerNum = m_Layers.size() - 2; layerNum > 0; --layerNum)
	//{
	//	Layer &hiddenLayer = m_Layers[layerNum];
	//	Layer &nextLayer = m_Layers[layerNum + 1];

	//	for (unsigned int n = 0; n < hiddenLayer.size(); ++n)
	//	{
	//		hiddenLayer[n].calcHiddenGradients(nextLayer);
	//	}
	//}

	//for all layers from outputs to first hidden layer: update connection weights
	for (unsigned int layerNum = m_Layers.size() - 1; layerNum > 0; --layerNum)
	{
		Layer &layer = m_Layers[layerNum];
		Layer &prevlayer = m_Layers[layerNum - 1];

		for (unsigned int n = 0; n < layer.size() - 1; ++n)
		{
			layer[n].updateInputWeights(prevlayer);
		}
	}

}


void Net::feedForward(const std::vector<double>& inputVals)
{
	//error check if size of inputlayer is correct
	assert(inputVals.size() == m_Layers[0].size() - 1);

	//assign the input values to input neurons
	for (unsigned int i = 0; i < inputVals.size(); ++i)
	{
		m_Layers[0][i].SetOutputValue(inputVals[i]);
	}

	//forward propagation = looping through each neuron -> ask to feed forward
	for (unsigned int layerNum = 1; layerNum < m_Layers.size(); ++layerNum)
	{
		Layer &pPrevLayer = m_Layers[layerNum - 1];
		for (unsigned int n = 0; n < m_Layers[layerNum].size() - 1; ++n)
		{
			m_Layers[layerNum][n].feedForward(pPrevLayer);
		}
	}

}

File: unity/Prog4/Assets/scripts/Net.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System.IO;
using System.Text;
//using Layer = System.Collections.Generic.List<Neuron>;

public class Net //: MonoBehaviour
{
    private List<List<Neuron>> m_Layers = new List<List<Neuron>>();

    public Net(List<int> topology)
    {
        Debug.Log("making net");
        string line;
        StreamReader thereader = new StreamReader("../prog4/Assets/resources/finalweights2.txt", Encoding.Default);


        int numLayers = topology.Count;
        for (int layerNum = 0; layerNum < numLayers; ++layerNum)
        {
            int numOutputs = 0;
            if (layerNum != topology.Count - 1)
                numOutputs = topology[layerNum + 1];

            var layer = new List<Neuron>();
            m_Layers.Add(layer);
            for (int neuronNum = 0; neuronNum <= topology[layerNum]; ++neuronNum)//1 bias neuron so 1 more loop
            {
                List<double> weights = new List<double>();
                for (int i = 0; i < 10; ++i)
                {
                    line = thereader.ReadLine();
                    double weight = 0;
                    if (line != null)
                    {
                        weight = double.Parse(line, System.Globalization.CultureInfo.InvariantCulture);
                    }
                    else
                    {
                        
                    }
                    weights.Add(weight);
                }
                

                m_Layers[m_Layers.Count - 1].Add(new Neuron(numOutputs, neuronNum, ref weights));//back gives most recent element
            }

            //force the bias neuron to output 1
            m_Layers[m_Layers.Count - 1][m_Layers[m_Layers.Count - 1].Count - 1].SetOutputValue(1.0);
        }
        thereader.Close();
    }

    public void FeedFoward(ref List<double> inputVals)
    {
        //set input vals
        for (int i = 0; i < inputVals.Count; ++i)
        {
            m_Layers[0][i].SetOutputValue(inputVals[i]);
        }

        //feed 
        for (int layerNum = 1; layerNum < m_Layers.Count; ++layerNum)
        {
            List<Neuron> pPrevLayer = m_Layers[layerNum - 1];
            for (int n = 0; n < m_Layers[layerNum].Count - 1; ++n)
            {
                m_Layers[layerNum][n].feedForward(ref pPrevLayer);
            }
        }
    }

    public void GetResults(ref List<double> resultsVals)
    {
    
        resultsVals.Clear();
    	for (int n = 0; n<m_Layers[m_Layers.Count - 1].Count - 1; ++n)
    	{
    		resultsVals.Add(m_Layers[m_Layers.Count - 1][n].GetOutputValue());
    	}
    }
}
